[Merged by Bors] - feat (AlgebraicGeometry/Modules): stalks of the sheaf `M^~`

jjaassoonn opened the pull request:
Construct the isomorphism of $R$-modules from the stalk of the sheaf $M^{\~}$ on $Spec(R)$ to the localization $M_x$ as an $R$-module, where $R$ is a commutative ring and $M$ is an $R$-module. 

## Main definition

* `ModuleCat.tilde.stalkIso`: The isomorphism of `R`-modules from the stalk of `M^~` at `x` to
the localization of `M` at the prime ideal corresponding to `x`.

## Technical note

To get the `R`-module structure on the stalks on `M^~`, we had to define
`ModuleCat.tildeInModuleCat`, which is `M^~` seen as sheaf of `R`-modules. We get it by
applying a forgetful functor to `ModuleCat.tilde M`.

Co-authored-by: Weihong Xu <weihong@vt.edu>
Co-authored-by: Sophie Morel <sophie.morel@ens-lyon.fr>
Co-authored-by: Amelia Livingston <al3717@ic.ac.uk>

This contribution was created as part of the AIM workshop "Formalizing algebraic geometry" in
June 2024.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #14254
- [x] depends on: #14231
- [x] depends on: #14809 

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jjaassoonn commented:
Should be compiling; sorry free; but need clean up

riccardobrasca commented:
Is this ready for review? If not, can you add "WIP"? Thanks!

smorel394 commented:
Hi Jujian, I did not realize that you had already PRed this branch! Anyway I tried to clean it up, put all the `M^~` in the same file in `AlgebraicGeometry/Modules`, added doc strings etc. If I messed something up, feel free to revert to an earlier commit.
Also, I got a strange error from CI last time I tried to build, it's complaining that two lemmas are not in `simp` normal form, but the expression it wants me to put on the LHS looks identical to the expression that is already there, so I had no idea what was going on and I gave up.

jjaassoonn commented:
> Hi Jujian, I did not realize that you had already PRed this branch! Anyway I tried to clean it up, put all the `M^~` in the same file in `AlgebraicGeometry/Modules`, added doc strings etc. If I messed something up, feel free to revert to an earlier commit. Also, I got a strange error from CI last time I tried to build, it's complaining that two lemmas are not in `simp` normal form, but the expression it wants me to put on the LHS looks identical to the expression that is already there, so I had no idea what was going on and I gave up. As far as I can tell, this is the last remaining CI problem.


Thanks for doing this! I have no idea either, so I just removed the `simp` attribute

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+      (TopCat.Presheaf.germ (tildeInModuleCat M) U x hx s) = (s.1 ⟨x, hx⟩ : _) :=
+  DFunLike.ext_iff.1 (germ_comp_stalkToFiberLinearMap M U x hx) s
+
+@[simp]
```
```suggestion
@[reassoc (attr := simp)]
```

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+
+@[simp]
+theorem toOpen_germ (U : Opens (PrimeSpectrum.Top R)) (x) (hx : x ∈ U) :
+    toOpen M U ≫ TopCat.Presheaf.germ (tildeInModuleCat M) U x hx = toStalk M x := by
```
```suggestion
    toOpen M U ≫ M.tildeInModuleCat.germ U x hx = toStalk M x := by
```

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+  rw [toStalk, Category.assoc, germ_comp_stalkToFiberLinearMap]; rfl
+
+theorem stalkToFiberLinearMap_toStalk (x : PrimeSpectrum.Top R) (m : M) :
+    (stalkToFiberLinearMap M x) ((toStalk M x) m) =
```
```suggestion
    (stalkToFiberLinearMap M x) (toStalk M x m) =
```

erdOne replied:
Does `elementwise` work here?

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+    rw [← res_apply M U V iVU s ⟨x, hxV⟩, ← hs, const_apply, localizationToStalk_mk]
+    exact (tildeInModuleCat M).germ_ext V hxV (homOfLE hg) iVU <| hs ▸ rfl
+  inv_hom_id := by ext x; exact x.induction_on (fun _ _ => by simp)
+
```
Can you provide an `IsLocalizedModule` instance?

jjaassoonn replied:
Thanks. Provided

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+
+@[simp]
+theorem germ_comp_stalkToFiberLinearMap (U : Opens (PrimeSpectrum.Top R)) (x) (hx : x ∈ U) :
+    TopCat.Presheaf.germ (tildeInModuleCat M) U x hx ≫ stalkToFiberLinearMap M x =
```
```suggestion
    (tildeInModuleCat M).germ U x hx ≫ stalkToFiberLinearMap M x =
```

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+prime ideal of `R` corresponding to `x`.
+-/
+noncomputable def stalkToFiberLinearMap (x : PrimeSpectrum.Top R) :
+    TopCat.Presheaf.stalk (tildeInModuleCat M) x ⟶
```
```suggestion
    (tildeInModuleCat M).stalk  x ⟶
```

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+implemented as a subtype of dependent functions to localizations at prime ideals, and evaluates
+the section on the point corresponding to a given prime ideal. -/
+def openToLocalization (U : Opens (PrimeSpectrum R)) (x : PrimeSpectrum R) (hx : x ∈ U) :
+    (tildeInModuleCat M).1.obj (op U) ⟶
```
```suggestion
    (tildeInModuleCat M).presheaf.obj (op U) ⟶
```

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+  rw [← toOpen_res M ⊤ U (homOfLE le_top : U ⟶ ⊤), Category.assoc, Presheaf.germ_res]; rfl
+
+theorem germ_toOpen (U : Opens (PrimeSpectrum.Top R)) (x) (hx : x ∈ U) (f : M) :
+    (M.tildeInModuleCat.germ U x hx) ((ModuleCat.Tilde.toOpen M U) f) = toStalk M x f := by
```
```suggestion
    M.tildeInModuleCat.germ U x hx (ModuleCat.Tilde.toOpen M U f) = toStalk M x f := by
```

erdOne replied:
Does this lemma come for free if you put `@[elementwise]` on `toOpen_germ`?

jjaassoonn replied:
indeed!

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,140 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+    (M.tildeInModuleCat.germ U x hx) ((ModuleCat.Tilde.toOpen M U) f) = toStalk M x f := by
+  rw [← toOpen_germ]; rfl
+
+@[simp]
```
```suggestion
@[reassoc (attr := simp)]
```

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,142 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+    exact (tildeInModuleCat M).germ_ext V hxV (homOfLE hg) iVU <| hs ▸ rfl
+  inv_hom_id := by ext x; exact x.induction_on (fun _ _ => by simp)
+
+instance (x : PrimeSpectrum.Top R) : IsLocalizedModule x.asIdeal.primeCompl (stalkIso M x).hom :=
```
No this means `M_x` is the localization of `M_x` at `x` right?

jjaassoonn replied:
Indeed

jjaassoonn replied:
now it should be okay

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,145 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+  inv_hom_id := by ext x; exact x.induction_on (fun _ _ => by simp)
+
+instance (x : PrimeSpectrum.Top R) :
+    IsLocalizedModule x.asIdeal.primeCompl (toOpen _ _ ≫ M.tildeInModuleCat.germ ⊤ x ⟨⟩) := by
```
The map is just `toStalk` right?

jjaassoonn replied:
Indeed. Thanks

erdOne commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -252,6 +254,145 @@ noncomputable def localizationToStalk (x : PrimeSpectrum.Top R) :
+        obtain ⟨c, hc⟩ := h2
+        exact LocalizedModule.mk_eq.mpr ⟨c, by simpa using hc.symm⟩⟩
+
+
```
```suggestion
```

erdOne reviewed (approved):
LGTM otherwise

erdOne reviewed (approved):
Thanks!
maintainer merge

joelriou commented:
Thanks!

bors merge
